### DISCUSSION TITLE
Solution for near-global atom use in atomFamily params

### DISCUSSION THREAD

**Discussion by awdyson** [facebookexperimental/Recoil#1162] [Q&A]
I'm curious if there's a way for `atomFamily`s to rely on an `atom` without needing to fetch/pass it in each consuming component.

Here's an example of a code pattern I find myself repeating all over:
```tsx
const brandId = useRecoilValue(brandStore.currentBrandId);
const products = useRecoilValue(productStore.productList({ brandId }));
```
`currentBrandId` is an `atom` because it's not uncommonly user-changed.
`productList` is an `atomFamily` to deal with things like data refreshes/updates.

We're already saving `currentBrandId` in `sessionStorage` as part of its set method.
It'd be nice to register a listener on `sessionStorage` to keep the logic together and smaller.

**Comment by rbaxter08** [comment]
Seems like you could accomplish this with a `selector` / `selectorFamily` https://recoiljs.org/docs/api-reference/utils/selectorFamily/

```ts
// import currentBrandId Atom
import { currentBrandId } from ...

const currentProductList= selector({
  key: 'current-product-list',
  get: ({get}) => {
    const brandId = get(currentBrandId)
    // logic from "productList" to return product list for the brandId
  }
});
```

**Reply by awdyson** [reply]
And if I have logic that refetches this data after certain actions?
`selector` and `selectorFamily` don't seem to have a way to update their own data (which makes sense).

**Reply by awdyson** [reply]
FWIW, my current approach works.
I'm just hoping there some simplification that I'm missing.

**Reply by rbaxter08** [reply]
I'm not 100% I follow (maybe a more complete example could help), but if you're saying your current solution updates after a certain data or action, then I assume that happens inside your `productStore.productList`, so putting that logic inside the `currentProductList` selector should do the same.

A selector will update whenever any of the atoms it depends on update.

**Reply by awdyson** [reply]
I'm saying that I've got a method like this:
```tsx
const brandId = useRecoilValue(brandStore.currentBrandId);
const [products, setProducts] = useRecoilState(productStore.productList({ brandId }));

// ...

async function addProduct(productData) {
  await api.postProduct(productData);
  const newProducts = await api.getProducts({ brandId });
  setProducts(newProducts);
}
```

`productList` only depends on `brandId`, but it can be updated by other factors.
That means that a selector[Family] isn't fitting, right?

**Reply by awdyson** [reply]
Am I understanding the "ReadWrite" `Selector` correctly?
That it can trigger `set` calls on `atom`s but can't update itself?

**Reply by rbaxter08** [reply]
Selectors can also define a set method.  So, if you have one atom to store the brandId and an atomFamily to store the productList, your code could look something like

```ts
import { currentBrandIdAtom } from ...
import { productListAtomFamily } from ...

const selectCurrentProductList= selector({
  key: 'current-product-list',
  get: ({get}) => {
    const brandId = get(currentBrandIdAtom);
    return get(productListAtom({ brandId }));
  },
  set: ({ get, set}, productList) => {
    const brandId = get(currentBrandIdAtom);
    set(productListAtomFamily({ brandId }), productList);
  }
});
```

```ts
const brandId = useRecoilValue(productStore.brandId);
const [products, setProducts] = useRecoilState(productStore.selectProductList());

// ...

async function addProduct(productData) {
  await api.postProduct(productData);
  const newProducts = await api.getProducts({ brandId });
  setProducts(newProducts);
}
```

**Reply by rbaxter08** [reply]
>That it can trigger set calls on atoms but can't update itself?

right, because selectors themselves don't store state, they represent derived state.  So, you don't use the selectors `set` to update itself, but to update other Atoms.

**Reply by awdyson** [reply]
Alright, I thought that might be the suggested solution.
I think I'll keep the redundancy of needing `currentBrandId` everywhere.
For this project, at least, it feels like "less abstraction in stores" > "less recoil hook repetition".

Thank you very much for the quick responses.
At the very least, you've helped me feel more confident in my understanding of this library :)

Also I guess, this can be considered a vote for adding something like `watchAtom` or `watchStorage` to `AtomOptions`.
If that's a terrible idea, feel free to just ignore it.

**Comment by drarmstr** [comment]
Thanks @rbaxter08   Also please see [this example](https://recoiljs.org/docs/api-reference/utils/atomFamily#scoped-atoms) in the docs for this pattern.
